fatfs: Update error code mapping

geky opened the pull request:
### Description

Continues https://github.com/ARMmbed/mbed-os/issues/6072, https://github.com/ARMmbed/mbed-os/pull/6120, https://github.com/ARMmbed/mbed-os/pull/6323

A lot of the error codes in fatfs were mapped incorrectly. This patch revisits the error code mapping to try to correct these mistakes.

Note: This is required before we can test LittleFS + FAT consistently

| ChanFS                  | was        | should be  | ChanFS description                                               |
| ----------------------- | ---------- | ---------- | ---------------------------------------------------------------- |
| FR_OK                   | 0          | 0          | Succeeded                                                        |
| FR_DISK_ERR             | EIO        | EIO        | A hard error occurred in the low level disk I/O layer            |
| FR_INT_ERR              | ~EBADF~    | -1         | Assertion failed                                                 |
| FR_NOT_READY            | EIO        | EIO        | The physical drive cannot work                                   |
| FR_NO_FILE              | ENOENT     | ENOENT     | Could not find the file                                          |
| FR_NO_PATH              | ~ENOENT~   | ENOTDIR    | Could not find the path                                          |
| FR_INVALID_NAME         | ~ENOENT~   | EINVAL     | The path name format is invalid                                  |
| FR_DENIED               | EACCES     | EACCES     | Access denied due to prohibited access or directory full         |
| FR_EXIST                | EEXIST     | EEXIST     | Access denied due to prohibited access                           |
| FR_INVALID_OBJECT       | ~EFAULT~   | EBADF      | The file/directory object is invalid                             |
| FR_WRITE_PROTECTED      | EACCES     | EACCES     | The physical drive is write protected                            |
| FR_INVALID_DRIVE        | ~ENOENT~   | ENODEV     | The logical drive number is invalid                              |
| FR_NOT_ENABLED          | ~ENXIO~    | ENODEV     | The volume has no work area                                      |
| FR_NO_FILESYSTEM        | ~ENOENT~   | EINVAL     | There is no valid FAT volume                                     |
| FR_MKFS_ABORTED         | ~EBADF~    | EIO        | The f_mkfs() aborted due to any problem                          |
| FR_TIMEOUT              | ~EBADF~    | ETIMEDOUT     | Could not get a grant to access the volume within defined period |
| FR_LOCKED               | ~EACCES~   | EBUSY      | The operation is rejected according to the file sharing policy   |
| FR_NOT_ENOUGH_CORE      | ENOMEM     | ENOMEM     | LFN working buffer could not be allocated                        |
| FR_TOO_MANY_OPEN_FILES  | ENFILE     | ENFILE     | Number of open files > FF_FS_LOCK                                |
| FR_INVALID_PARAMETER    | ~ENOEXEC~  | EINVAL     | Given parameter is invalid                                       |

cc @kegilbert, @deepikabhavnani, @kjbracey-arm, @JanneKiiskila, @teetak01 
intended for minor release
should resolve https://github.com/ARMmbed/mbed-os/issues/6072, https://github.com/ARMmbed/mbed-os/issues/5653

### Pull request type

<!-- 
    Required
    Please tick one of the following types 
-->

- [ ] Fix
- [ ] Refactor
- [ ] New target
- [ ] Feature
- [x] Breaking change

geky committed:
fatfs: Update error code mapping

A lot of the error codes in fatfs were mapped incorrectly. This patch
revisits the error code mapping to try to correct these mistakes.

geky committed:
fatfs: Fixed ENOTEMPTY error for removing a directory

geky committed:
retarget: Added include errno.h

Unfortunately, some toolchains don't define the same errno codes that
are used fairly consistently on Linux based platforms, which means they
also don't match the errno codes used in the retarget layer.

If a user includes errno.h after mbed.h, the errno codes can be
redefined incorrectly.

Adding an include of errno.h in mbed.h forces the order to be fixed.

cmonr commented:
@geky @JanneKiiskila Is it safe to say that this is now unblocked and can proceed into 5.9?

geky commented:
@teetak01, can this be progressed?

teetak01 commented:
http://github.com/armmbed/mbed-cloud-client-example should be compatible. At least it was working one month ago still. The required fix on Client-side was already part of 1.3.1 release.

teetak01 commented:
Are you running any Client CI-verification as part of the PRs? Currently the checks look to be a bit light-weight.

geky commented:
> Client CI-verification

I'm not really sure, is this what pr-head is?
https://jenkins-internal.mbed.com/job/ARMmbed/job/mbed-os/job/PR-6336/3/display/redirect

teetak01 commented:
It only builds Mbed OS and unrelated application, not mbed-cloud-client-example.

cc: @OPpuolitaival

geky commented:
Ah, then there's no client verification running.

cmonr commented:
I think the only client verification we're running is either through pr-head/cli-app and/or whatever examples may be enabled in tools/test/examples/examples.json

Adding the cloud-client example to examples.json might be a good first step.

cmonr commented:
@geky I think this would be good to progress, but I would suggest starting with a rebase. pr-head failed with an odd error, indicating the K64F couldn't work in Tickless mode.

cmonr commented:
@geky Hold off on that rebase. Looks like something is amis with pr-head.

JanneKiiskila commented:
Client team has  tests running against the Mbed OS master and I think @OPpuolitaival  also has now some tests on the Mbed OS with the Client as well? Just recently added, like a few days ago. So from my point of view this can be merged as soon you get the PR in OK shape.

cmonr commented:
@geky Take a look at the pr-head/cli-app failure. On a quick check, I can't tell what's wrong with the build, and/or if it's related to the PR. It would be interesting if it's caused by the change in `platform/mbed_retarget.h`.

@OPpuolitaival @SeppoTakalo Could we ask for your help here? There were one or two other PRs that failed pr-head with `[Error] main.cpp@17,2: #error [NOT_SUPPORTED] Tickless mode not supported for this target.`. I'm wondering if these PRs simply need to be rebased, since I haven't noticed any newer PRs failing.

SeppoTakalo commented:
I can see some build failures in mbed-os-cliapp builds:
```
[Error] errno.h@5,13: conflicting declaration 'typedef int error_t'
```

geky commented:
Oh that's not good, errno.h and cliapp are incompatible

geky committed:
retarget: Added workaround for conflicting definitions of error_t

The errno.h header file defines the type error_t, unfortunately this
is a common type name that may be defined in user code. For at least GCC
we can work around this by telling errno that the error_t is already
defined.

geky commented:
Rebased and added a workaround for the `error_t` conflict

cmonr commented:
Will restart CI later today. Giving PRs targeted for 5.8.5 a chance to complete.

adbridge commented:
@kjbracey-arm  @cmonr  @0xc0170  Are you happy with the updates ? Can you please approve if so?

cmonr commented:
My bad. Forgot I was a reviewer ^_^

/morph build

0xc0170 commented:
> I would recommend to keep this PR open until IOTCLT-2533 is done.

@teetak01 I checked and its closed. All good for this PR to be approved?

teetak01 reviewed (commented):
should be fine. I only tested the original PR.
